# MovieLens 25M - ML Recommendation System 

## Data Preparation Notebook

This notebook is based on the previous one (01_EDA_MovieLens_25M.ipynb) and its objective is to prepare the data for proper ML model training for a recommendation system for movies. We will be taking the strategic choice of which data sources to feed into the recommender, because each file in MovieLens 25M serves a different purpose, and the “best” one depends on the type of model we want to build. 

In the previous notebook we've seen that the MovieLens 25M Dataset is composed of several files:

    - ratings.csv → the primary dataset for training collaborative filtering models (user-based or item-based): it contains explicit user feedback (0.5–5 stars)

    - movies.csv → a supporting file for hybrid models containing metadata (title, genres); it will be useful for content-based features (genre embeddings, one-hot vectors) and shall help with cold-start items (new movies with few ratings)

    - genome-tags.csv + genome-scores.csv → Semantic features: they provide a dense tag–movie relevance matrix (0–1 scores) and will be excellent for hybrid recommenders: we will combine collaborative signals with semantic descriptors

    - tags.csv → raw user-contributed tags, they are rather noisy (might be valuable for analyzing user language and preferences, but less useful or reliable for ML than genome tags)
    
    - links.csv → external IDs (IMDb, TMDb): not really needed for core training, but useful if we want to enrich with external metadata (cast, crew, posters).

## Data strategy for ML Recommendation System

For further work on the ML model, we will focus on the three
 of the above described files:

    - ratings.csv - the 'backbone' of the whole dataset containing explicit feedback from users

    - movies.csv - the metadata	containing the genre and title features for content-based signals

    - genome-scores.csv - the semantic features	with dense tag relevance scores for hybrid modeling

We could link them all thanks to the 'movieId' column as a join key, but their size and internal structure will not allow us to do this because of memory limitations. Therefore we will produce 2 merged files which we consider as the right strategy for building our recommendation model because it will allow us to work on:

    Collaborative filtering → as a starting point for training a baseline model        

    Hybrid (collaborative + content based model) → for model enrichment and optimization

In the last part of this notebook we will proceed with splitting the entire dataframe (ratings_movies) with the following proportions: 70/20/10. The split will be temporal (based on timestamp) from the oldest to the most recent ratings and will serve the following objectives respectively:

- ML model training (70% of the data)
- ML model testing (20% of the data
- Continuous buffer (10% of the data): for continuous training after deployment)


It is important to mention at this stage that the 'continuous buffer' will simulate the 'new' data ingestion so that the model can be properly monitored and retrained if needed.



In [1]:
# Importing libraries for data manipulation, visualization, and analysis

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
# Loading files
ratings = pd.read_csv("../raw/ratings.csv")
movies = pd.read_csv("../raw/movies.csv")
genome_scores = pd.read_csv("../raw/genome-scores.csv")
genome_tags = pd.read_csv("../raw/genome-tags.csv")  # Optional for tag names (in case needed)


In [3]:
# Saving movies_base for future use
movies_base = movies.copy()

In [4]:
# Displaying the first few rows of each dataframe to remind their structure
display(ratings.head())
display(movies.head())
display(genome_scores.head())
display(genome_tags.head())  # Optional display


userId  movieId  rating   timestamp
0       1      296     5.0  1147880044
1       1      306     3.5  1147868817
2       1      307     5.0  1147868828
3       1      665     5.0  1147878820
4       1      899     3.5  1147868510

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

movieId  tagId  relevance
0        1      1    0.02875
1        1      2    0.02375
2        1      3    0.06250
3        1      4    0.07575
4        1      5    0.14075

tagId           tag
0      1           007
1      2  007 (series)
2      3  18th century
3      4         1920s
4      5         1930s

In [5]:
# Focusing on ratings first
ratings['datetime'] = pd.to_datetime(ratings['timestamp'], unit='s') # Convert timestamp to datetime
ratings['year'] = ratings['datetime'].dt.year # Extract year from datetime
ratings['liked'] = (ratings['rating'] >= 3.5).astype(int) # Create binary liked column

In [6]:
# Displaying the first few rows of ratings to check new columns
display(ratings.head())

userId  movieId  rating   timestamp            datetime  year  liked
0       1      296     5.0  1147880044 2006-05-17 15:34:04  2006      1
1       1      306     3.5  1147868817 2006-05-17 12:26:57  2006      1
2       1      307     5.0  1147868828 2006-05-17 12:27:08  2006      1
3       1      665     5.0  1147878820 2006-05-17 15:13:40  2006      1
4       1      899     3.5  1147868510 2006-05-17 12:21:50  2006      1

In [7]:
# Focusing on movies next - exploding genres so each row contains only one genre
# Split genres into list
movies['genres'] = movies['genres'].apply(lambda g: g.split('|') if isinstance(g, str) else [])
# Explode so each movie-genre combination is a separate row
movies = movies.explode('genres').reset_index(drop=True)

In [8]:
# Checking the movies dataframe after genre expansion
display(movies.head())

movieId             title     genres
0        1  Toy Story (1995)  Adventure
1        1  Toy Story (1995)  Animation
2        1  Toy Story (1995)   Children
3        1  Toy Story (1995)     Comedy
4        1  Toy Story (1995)    Fantasy

In [9]:
# Checking the shape of the movies dataframe (after exploding)
movies.shape

(112307, 3)

In [10]:
# Focusing on genome scores by aggregating them into a wide format
genome_wide = genome_scores.pivot(index="movieId", columns="tagId", values="relevance").astype('float32')

This pivot line is reshaping the genome scores from “long” (many rows per movie) into “wide” (one row per movie, many tag columns). This would be a standard step in feature engineering so that we can treat tags as numeric features. The pivot produces a matrix, where each row is a movie’s “semantic fingerprint” across all tags ( it is possible in further part to reduce the dimensionality as the number of columns is significant).

In [11]:
# Checking the genome_wide dataframe
display(genome_wide.head())

tagId       1        2        3        4        5        6        7     \
movieId                                                                  
1        0.02875  0.02375  0.06250  0.07575  0.14075  0.14675  0.06350   
2        0.04125  0.04050  0.06275  0.08275  0.09100  0.06125  0.06925   
3        0.04675  0.05550  0.02925  0.08700  0.04750  0.04775  0.04600   
4        0.03425  0.03800  0.04050  0.03100  0.06500  0.03575  0.02900   
5        0.04300  0.05325  0.03800  0.04100  0.05400  0.06725  0.02775   

tagId       8       9        10    ...     1119     1120     1121     1122  \
movieId                            ...                                       
1        0.20375  0.2020  0.03075  ...  0.04050  0.01425  0.03050  0.03500   
2        0.09600  0.0765  0.05250  ...  0.05250  0.01575  0.01250  0.02000   
3        0.14275  0.0285  0.03875  ...  0.06275  0.01950  0.02225  0.02300   
4        0.08650  0.0320  0.03150  ...  0.05325  0.02800  0.01675  0.03875   
5        0.07650  0.0215  0.02975  ...  0.05350  0.02050  0.01425  0.02550   

tagId       1123     1124     1125     1126     1127     1128  
movieId                                                        
1        0.14125  0.05775  0.03900  0.02975  0.08475  0.02200  
2        0.12225  0.03275  0.02100  0.01100  0.10525  0.01975  
3        0.12200  0.03475  0.01700  0.01800  0.09100  0.01775  
4        0.18200  0.07050  0.01625  0.01425  0.08850  0.01500  
5        0.19225  0.02675  0.01625  0.01300  0.08700  0.01600  

[5 rows x 1128 columns]

In [12]:
# Checking the pivoted dataframe shape
genome_wide.shape

(13816, 1128)

In [13]:
# Handle missing genome scores (movies without genome tags)
# Option: Fill NaN with 0 (assumes no tag relevance if not rated)
genome_wide = genome_wide.fillna(0)

print(f"Genome wide shape after filling NaNs: {genome_wide.shape}")
print(f"Number of unique movies in genome_wide: {len(genome_wide)}")


Genome wide shape after filling NaNs: (13816, 1128)
Number of unique movies in genome_wide: 13816


In [14]:
# Prepare genome_wide for merge with movies - reset index so movieId is a column
genome_wide_reset = genome_wide.reset_index()
display(genome_wide_reset.head())


tagId  movieId        1        2        3        4        5        6        7  \
0            1  0.02875  0.02375  0.06250  0.07575  0.14075  0.14675  0.06350   
1            2  0.04125  0.04050  0.06275  0.08275  0.09100  0.06125  0.06925   
2            3  0.04675  0.05550  0.02925  0.08700  0.04750  0.04775  0.04600   
3            4  0.03425  0.03800  0.04050  0.03100  0.06500  0.03575  0.02900   
4            5  0.04300  0.05325  0.03800  0.04100  0.05400  0.06725  0.02775   

tagId        8       9  ...     1119     1120     1121     1122     1123  \
0      0.20375  0.2020  ...  0.04050  0.01425  0.03050  0.03500  0.14125   
1      0.09600  0.0765  ...  0.05250  0.01575  0.01250  0.02000  0.12225   
2      0.14275  0.0285  ...  0.06275  0.01950  0.02225  0.02300  0.12200   
3      0.08650  0.0320  ...  0.05325  0.02800  0.01675  0.03875  0.18200   
4      0.07650  0.0215  ...  0.05350  0.02050  0.01425  0.02550  0.19225   

tagId     1124     1125     1126     1127     1128  
0      0.05775  0.03900  0.02975  0.08475  0.02200  
1      0.03275  0.02100  0.01100  0.10525  0.01975  
2      0.03475  0.01700  0.01800  0.09100  0.01775  
3      0.07050  0.01625  0.01425  0.08850  0.01500  
4      0.02675  0.01625  0.01300  0.08700  0.01600  

[5 rows x 1129 columns]

## Merge strategy
- Join `ratings` with `movies` (not exploded).
- Keep genome tag features at the item level: `movie_features = movies + genome_wide`.
- We will avoid merging genome_wide onto every rating row (25M x 1,128 tag cols ≈ >110 GB in float32).
- When training, join `movie_features` on `movieId` in batches or load as a separate lookup.

In [15]:
# Join ratings with the base movies file (non-exploded) to keep ~25M rows
ratings_movies = ratings.merge(movies_base, on="movieId", how="left")
print(f"ratings_movies shape: {ratings_movies.shape}")

ratings_movies shape: (25000095, 9)


In [16]:
display(ratings_movies.head())

userId  movieId  rating   timestamp            datetime  year  liked  \
0       1      296     5.0  1147880044 2006-05-17 15:34:04  2006      1   
1       1      306     3.5  1147868817 2006-05-17 12:26:57  2006      1   
2       1      307     5.0  1147868828 2006-05-17 12:27:08  2006      1   
3       1      665     5.0  1147878820 2006-05-17 15:13:40  2006      1   
4       1      899     3.5  1147868510 2006-05-17 12:21:50  2006      1   

                                              title  \
0                               Pulp Fiction (1994)   
1  Three Colors: Red (Trois couleurs: Rouge) (1994)   
2  Three Colors: Blue (Trois couleurs: Bleu) (1993)   
3                                Underground (1995)   
4                        Singin' in the Rain (1952)   

                        genres  
0  Comedy|Crime|Drama|Thriller  
1                        Drama  
2                        Drama  
3             Comedy|Drama|War  
4       Comedy|Musical|Romance

In [17]:
# Build item-level features (movies + genome_wide) separately to avoid blowing up memory
movie_features = movies.merge(genome_wide_reset, on="movieId", how="left")

In [18]:
display(movie_features.head())

movieId             title     genres        1        2       3        4  \
0        1  Toy Story (1995)  Adventure  0.02875  0.02375  0.0625  0.07575   
1        1  Toy Story (1995)  Animation  0.02875  0.02375  0.0625  0.07575   
2        1  Toy Story (1995)   Children  0.02875  0.02375  0.0625  0.07575   
3        1  Toy Story (1995)     Comedy  0.02875  0.02375  0.0625  0.07575   
4        1  Toy Story (1995)    Fantasy  0.02875  0.02375  0.0625  0.07575   

         5        6       7  ...    1119     1120    1121   1122     1123  \
0  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
1  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
2  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
3  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
4  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   

      1124   1125     1126     1127   1128  
0  0.05775  0.039  0.02975  0.08475  0.022  
1  0.05775  0.039  0.02975  0.08475  0.022  
2  0.05775  0.039  0.02975  0.08475  0.022  
3  0.05775  0.039  0.02975  0.08475  0.022  
4  0.05775  0.039  0.02975  0.08475  0.022  

[5 rows x 1131 columns]

In [19]:
# Fill missing genome scores for movies without tags, then inspect shapes
if not genome_wide_reset.empty:
    genome_tag_cols = genome_wide_reset.columns.drop('movieId')
    movie_features[genome_tag_cols] = movie_features[genome_tag_cols].fillna(0)

print(f"ratings_movies shape: {ratings_movies.shape}")
print(f"movie_features shape: {movie_features.shape}")

display(movie_features.head())

ratings_movies shape: (25000095, 9)
movie_features shape: (112307, 1131)


movieId             title     genres        1        2       3        4  \
0        1  Toy Story (1995)  Adventure  0.02875  0.02375  0.0625  0.07575   
1        1  Toy Story (1995)  Animation  0.02875  0.02375  0.0625  0.07575   
2        1  Toy Story (1995)   Children  0.02875  0.02375  0.0625  0.07575   
3        1  Toy Story (1995)     Comedy  0.02875  0.02375  0.0625  0.07575   
4        1  Toy Story (1995)    Fantasy  0.02875  0.02375  0.0625  0.07575   

         5        6       7  ...    1119     1120    1121   1122     1123  \
0  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
1  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
2  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
3  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   
4  0.14075  0.14675  0.0635  ...  0.0405  0.01425  0.0305  0.035  0.14125   

      1124   1125     1126     1127   1128  
0  0.05775  0.039  0.02975  0.08475  0.022  
1  0.05775  0.039  0.02975  0.08475  0.022  
2  0.05775  0.039  0.02975  0.08475  0.022  
3  0.05775  0.039  0.02975  0.08475  0.022  
4  0.05775  0.039  0.02975  0.08475  0.022  

[5 rows x 1131 columns]

### Single-row item features (preferred for hybrid)
We'll create `movie_features_uni` with one row per `movieId` by using genre one-hot (multi-hot) from the base movies and merging genome tags. This is smaller and cleaner for training-time lookups.

In [20]:
# Build per-movie item features: genres one-hot (multi-hot) + genome tags
# 1) Genres as one-hot from base movies (no explosion)
genre_dummies = movies_base['genres'].str.get_dummies(sep='|').astype('int8')
movie_features_uni = movies_base[['movieId', 'title']].join(genre_dummies)

# 2) Merge genome tag features (wide) and fill NaNs
movie_features_uni = movie_features_uni.merge(genome_wide_reset, on='movieId', how='left')
if not genome_wide_reset.empty:
    genome_cols = genome_wide_reset.columns.drop('movieId')
    # reduce memory: keep genome features as float32
    movie_features_uni[genome_cols] = movie_features_uni[genome_cols].fillna(0).astype('float32')

print(f"movie_features (exploded) shape: {movie_features.shape}")
print(f"movie_features_uni (1 row/movie) shape: {movie_features_uni.shape}")

display(movie_features_uni.head())

movie_features (exploded) shape: (112307, 1131)
movie_features_uni (1 row/movie) shape: (62423, 1150)


movieId                               title  (no genres listed)  Action  \
0        1                    Toy Story (1995)                   0       0   
1        2                      Jumanji (1995)                   0       0   
2        3             Grumpier Old Men (1995)                   0       0   
3        4            Waiting to Exhale (1995)                   0       0   
4        5  Father of the Bride Part II (1995)                   0       0   

   Adventure  Animation  Children  Comedy  Crime  Documentary  ...     1119  \
0          1          1         1       1      0            0  ...  0.04050   
1          1          0         1       0      0            0  ...  0.05250   
2          0          0         0       1      0            0  ...  0.06275   
3          0          0         0       1      0            0  ...  0.05325   
4          0          0         0       1      0            0  ...  0.05350   

      1120     1121     1122     1123     1124     1125     1126     1127  \
0  0.01425  0.03050  0.03500  0.14125  0.05775  0.03900  0.02975  0.08475   
1  0.01575  0.01250  0.02000  0.12225  0.03275  0.02100  0.01100  0.10525   
2  0.01950  0.02225  0.02300  0.12200  0.03475  0.01700  0.01800  0.09100   
3  0.02800  0.01675  0.03875  0.18200  0.07050  0.01625  0.01425  0.08850   
4  0.02050  0.01425  0.02550  0.19225  0.02675  0.01625  0.01300  0.08700   

      1128  
0  0.02200  
1  0.01975  
2  0.01775  
3  0.01500  
4  0.01600  

[5 rows x 1150 columns]

### Dimensionality reduction (PCA) for genome tags
With the above merge strategu, we obtained a large dataframe 'movie_features_uni' of the following shape: (62423, 1150). It will be useful for ML training to reduce the number of columns using PCA: we will project the ~1.1k genome-tag features into a compact embedding (64 dims). The genome scores are 0–1, so we will have to center features before PCA. This yields smaller per-movie vectors while retaining most variance.

In [21]:
# PCA embeddings for genome tags
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

# Choose genome columns directly from the wide table reference
genome_cols = list(genome_wide_reset.columns.drop('movieId')) if 'genome_wide_reset' in globals() else []

EMB_DIM = 64

if genome_cols:
    # Center features (no variance scaling), keep float32 to save memory
    scaler = StandardScaler(with_mean=True, with_std=False)
    X_centered = scaler.fit_transform(movie_features_uni[genome_cols].astype('float32'))

    pca = PCA(n_components=EMB_DIM, svd_solver='randomized', random_state=42)
    Z = pca.fit_transform(X_centered).astype('float32')

    movie_embeddings = pd.DataFrame(Z, columns=[f'g_emb_{i}' for i in range(Z.shape[1])])
    movie_embeddings.insert(0, 'movieId', movie_features_uni['movieId'].values)

    var_expl = float(pca.explained_variance_ratio_.sum())
    print(f"PCA embeddings shape: {movie_embeddings.shape}; explained variance: {var_expl:.3f}")

    # Build a compact item-features table: keep genres + embeddings (drop raw genome tag columns)
    genre_cols = [c for c in movie_features_uni.columns if c not in ['movieId','title'] and c not in set(genome_cols)]
    movie_features_small = movie_features_uni[['movieId','title'] + genre_cols].merge(movie_embeddings, on='movieId', how='left')

    print(f"movie_features_small shape: {movie_features_small.shape}")
    display(movie_features_small.head())
else:
    print("No genome columns found for PCA.")

PCA embeddings shape: (62423, 65); explained variance: 0.869
movie_features_small shape: (62423, 86)


movieId                               title  (no genres listed)  Action  \
0        1                    Toy Story (1995)                   0       0   
1        2                      Jumanji (1995)                   0       0   
2        3             Grumpier Old Men (1995)                   0       0   
3        4            Waiting to Exhale (1995)                   0       0   
4        5  Father of the Bride Part II (1995)                   0       0   

   Adventure  Animation  Children  Comedy  Crime  Documentary  ...  g_emb_54  \
0          1          1         1       1      0            0  ... -0.049991   
1          1          0         1       0      0            0  ...  0.080302   
2          0          0         0       1      0            0  ... -0.003513   
3          0          0         0       1      0            0  ... -0.316550   
4          0          0         0       1      0            0  ...  0.262982   

   g_emb_55  g_emb_56  g_emb_57  g_emb_58  g_emb_59  g_emb_60  g_emb_61  \
0  0.101698 -0.245650  0.004215 -0.133796  0.326354  0.073053  0.151898   
1  0.077730 -0.001964 -0.169890 -0.328397 -0.155135 -0.324006  0.229122   
2 -0.255320 -0.047668  0.043896  0.025104 -0.006311  0.120793  0.128007   
3  0.234972  0.107891  0.136243  0.067892  0.319507  0.008702  0.139247   
4 -0.265471  0.022237  0.152904  0.167361  0.148734  0.113982  0.046446   

   g_emb_62  g_emb_63  
0 -0.005448 -0.211587  
1  0.065407  0.075267  
2 -0.087302 -0.039096  
3 -0.150832  0.039255  
4 -0.158512  0.106765  

[5 rows x 86 columns]

In [22]:
movie_features_small.head()

movieId                               title  (no genres listed)  Action  \
0        1                    Toy Story (1995)                   0       0   
1        2                      Jumanji (1995)                   0       0   
2        3             Grumpier Old Men (1995)                   0       0   
3        4            Waiting to Exhale (1995)                   0       0   
4        5  Father of the Bride Part II (1995)                   0       0   

   Adventure  Animation  Children  Comedy  Crime  Documentary  ...  g_emb_54  \
0          1          1         1       1      0            0  ... -0.049991   
1          1          0         1       0      0            0  ...  0.080302   
2          0          0         0       1      0            0  ... -0.003513   
3          0          0         0       1      0            0  ... -0.316550   
4          0          0         0       1      0            0  ...  0.262982   

   g_emb_55  g_emb_56  g_emb_57  g_emb_58  g_emb_59  g_emb_60  g_emb_61  \
0  0.101698 -0.245650  0.004215 -0.133796  0.326354  0.073053  0.151898   
1  0.077730 -0.001964 -0.169890 -0.328397 -0.155135 -0.324006  0.229122   
2 -0.255320 -0.047668  0.043896  0.025104 -0.006311  0.120793  0.128007   
3  0.234972  0.107891  0.136243  0.067892  0.319507  0.008702  0.139247   
4 -0.265471  0.022237  0.152904  0.167361  0.148734  0.113982  0.046446   

   g_emb_62  g_emb_63  
0 -0.005448 -0.211587  
1  0.065407  0.075267  
2 -0.087302 -0.039096  
3 -0.150832  0.039255  
4 -0.158512  0.106765  

[5 rows x 86 columns]

## Interpretation

After the dimensionality reduction with PCA, we obtained a much smaller number of columns in the matrix: instead of 1150 columns we now have 86, where the number of rows stays the same: 62423 (this is the fixed number of all movies in the catalogue: each movie corresponds to one row and is explained as a dense vector composed of genres columns and Principal Components obtained from the PCA step). 

With this done, we can now prepared the temporal split of the data for training, testing and continuous buffer for the ML model for recommendation system. In order to avoid any random splitting and potential data leakage, we will use the timestamp a splitting criterion.

### Save artifacts and create 70/20/10 temporal split
We persist compact artifacts (Parquet) and split `ratings_movies` by `datetime` chronologically: oldest 70% → train, next 20% → test, most recent 10% → buffer.

In [29]:
# Save artifacts and perform temporal split
import os

out_dir = "../prepared"
os.makedirs(out_dir, exist_ok=True)

In [30]:
# Save main artifacts (use Parquet if available; fallback to CSV)
def _save(df, path_parquet, path_csv):
    try:
        df.to_parquet(path_parquet, index=False)
        print(f"Saved: {path_parquet}")
    except Exception as e:
        print(f"Parquet save failed ({e}); saving CSV instead.")
        df.to_csv(path_csv, index=False)
        print(f"Saved: {path_csv}")

_save(ratings_movies, os.path.join(out_dir, "ratings_movies.parquet"), os.path.join(out_dir, "ratings_movies.csv"))
_save(movie_features_uni, os.path.join(out_dir, "movie_features_uni.parquet"), os.path.join(out_dir, "movie_features_uni.csv"))
if 'movie_embeddings' in globals():
    _save(movie_embeddings, os.path.join(out_dir, "movie_embeddings.parquet"), os.path.join(out_dir, "movie_embeddings.csv"))
if 'movie_features_small' in globals():
    _save(movie_features_small, os.path.join(out_dir, "movie_features_small.parquet"), os.path.join(out_dir, "movie_features_small.csv"))


Saved: ../prepared\ratings_movies.parquet
Saved: ../prepared\movie_features_uni.parquet
Saved: ../prepared\movie_embeddings.parquet
Saved: ../prepared\movie_features_uni.parquet
Saved: ../prepared\movie_embeddings.parquet
Saved: ../prepared\movie_features_small.parquet
Saved: ../prepared\movie_features_small.parquet


In [31]:
# Temporal split 70/20/10 by datetime (global chronological order)
ratings_sorted = ratings_movies.sort_values('datetime')
N = len(ratings_sorted)
train_end = int(N * 0.70)
test_end = int(N * 0.90)

ratings_train = ratings_sorted.iloc[:train_end]
ratings_test = ratings_sorted.iloc[train_end:test_end]
ratings_buffer = ratings_sorted.iloc[test_end:]

print("Split sizes:", len(ratings_train), len(ratings_test), len(ratings_buffer))
print("Date ranges:")
print("train:", ratings_train['datetime'].min(), "→", ratings_train['datetime'].max())
print("test:", ratings_test['datetime'].min(), "→", ratings_test['datetime'].max())
print("buffer:", ratings_buffer['datetime'].min(), "→", ratings_buffer['datetime'].max())

Split sizes: 17500066 5000019 2500010
Date ranges:
train: 1995-01-09 11:46:49 → 2015-01-22 16:42:53
test: 2015-01-22 16:43:05 → 2018-01-04 02:44:50
buffer: 2018-01-04 02:45:08 → 2019-11-21 09:15:03


In [32]:
# Save splits
_save(ratings_train, os.path.join(out_dir, "ratings_train.parquet"), os.path.join(out_dir, "ratings_train.csv"))
_save(ratings_test, os.path.join(out_dir, "ratings_test.parquet"), os.path.join(out_dir, "ratings_test.csv"))
_save(ratings_buffer, os.path.join(out_dir, "ratings_buffer.parquet"), os.path.join(out_dir, "ratings_buffer.csv"))

Saved: ../prepared\ratings_train.parquet
Saved: ../prepared\ratings_test.parquet
Saved: ../prepared\ratings_test.parquet
Saved: ../prepared\ratings_buffer.parquet
Saved: ../prepared\ratings_buffer.parquet


## Quick Preview of Saved Artifacts
Below previews read only a few rows from Parquet/CSV to avoid loading full datasets into memory.

In [33]:
# Preview saved files: read small heads safely (Parquet/CSV)
# Parquet helpers
try:
    import pyarrow.parquet as pq
except Exception:
    pq = None

In [34]:
# Row-limited Parquet/CSV heads without full file load

def parquet_head(path, n=5, columns=None):
    if pq is None:
        print(f"pyarrow.parquet not available; skipping Parquet head for {path}")
        return None
    try:
        pf = pq.ParquetFile(path)
        dfs = []
        rows_collected = 0
        for rg in range(pf.num_row_groups):
            table = pf.read_row_group(rg, columns=columns)
            if table.num_rows == 0:
                continue
            take = min(n - rows_collected, table.num_rows)
            if take <= 0:
                break
            dfs.append(table.slice(0, take).to_pandas())
            rows_collected += take
            if rows_collected >= n:
                break
        if rows_collected == 0:
            # Empty file or no rows read; return empty frame with schema columns
            cols = columns if columns is not None else pf.schema.names
            return pd.DataFrame(columns=cols)
        return pd.concat(dfs, ignore_index=True)
    except Exception as e:
        print(f"Parquet head failed for {path}: {e}")
        return None


def csv_head(path, n=5, usecols=None):
    try:
        return pd.read_csv(path, nrows=n, usecols=usecols)
    except Exception as e:
        print(f"CSV head failed for {path}: {e}")
        return None

pairs = [
    ("ratings_train",   "../prepared/ratings_train.parquet",   "../prepared/ratings_train.csv",   ["userId","movieId","rating","datetime","title","genres"]),
    ("ratings_test",    "../prepared/ratings_test.parquet",    "../prepared/ratings_test.csv",    ["userId","movieId","rating","datetime","title","genres"]),
    ("ratings_buffer",  "../prepared/ratings_buffer.parquet",  "../prepared/ratings_buffer.csv",  ["userId","movieId","rating","datetime","title","genres"]),
    ("movie_features_small", "../prepared/movie_features_small.parquet", "../prepared/movie_features_small.csv", None),
    ("movie_features_uni",   "../prepared/movie_features_uni.parquet",   "../prepared/movie_features_uni.csv",   None),
    ("movie_embeddings",     "../prepared/movie_embeddings.parquet",     "../prepared/movie_embeddings.csv",     None),
]

for name, p_parq, p_csv, cols in pairs:
    print("\n===", name, "===")
    df = None
    src = None

    if os.path.exists(p_parq):
        df = parquet_head(p_parq, n=5, columns=cols)
        src = p_parq
    if df is None and os.path.exists(p_csv):
        df = csv_head(p_csv, n=5, usecols=cols)
        src = p_csv

    if df is not None:
        print(f"source: {src}")
        display(df)
    else:
        print(f"No preview available (files missing or reader not available): {p_parq} | {p_csv}")

# Additionally, show a compact column sample for movie_features_small if present
mfs_parq = "../prepared/movie_features_small.parquet"
if os.path.exists(mfs_parq) and pq is not None:
    try:
        sample_cols = ["movieId","title","g_emb_0","g_emb_1"]
        head_df = parquet_head(mfs_parq, n=5, columns=sample_cols)
        if head_df is not None:
            print("\nmovie_features_small sample columns:")
            display(head_df)
    except Exception as e:
        print(f"Extra sample failed: {e}")


=== ratings_train ===
source: ../prepared/ratings_train.parquet


userId  movieId  rating            datetime                        title  \
0    2262       21     3.0 1995-01-09 11:46:49            Get Shorty (1995)   
1    2262     1079     3.0 1995-01-09 11:46:49  Fish Called Wanda, A (1988)   
2    2262       47     5.0 1995-01-09 11:46:49  Seven (a.k.a. Se7en) (1995)   
3  102689        1     4.0 1996-01-29 00:00:00             Toy Story (1995)   
4  102689       39     5.0 1996-01-29 00:00:00              Clueless (1995)   

                                        genres  
0                        Comedy|Crime|Thriller  
1                                 Comedy|Crime  
2                             Mystery|Thriller  
3  Adventure|Animation|Children|Comedy|Fantasy  
4                               Comedy|Romance


=== ratings_test ===
source: ../prepared/ratings_test.parquet


userId  movieId  rating            datetime  \
0   84815    48394     4.0 2015-01-22 16:43:05   
1   84815    76251     4.0 2015-01-22 16:43:08   
2   84815    60684     5.0 2015-01-22 16:43:29   
3  133589    88810     3.5 2015-01-22 16:43:43   
4  129879    34405     4.0 2015-01-22 16:43:48   

                                              title  \
0  Pan's Labyrinth (Laberinto del fauno, El) (2006)   
1                                   Kick-Ass (2010)   
2                                   Watchmen (2009)   
3                                  Help, The (2011)   
4                                   Serenity (2005)   

                                      genres  
0                     Drama|Fantasy|Thriller  
1                              Action|Comedy  
2  Action|Drama|Mystery|Sci-Fi|Thriller|IMAX  
3                                      Drama  
4                    Action|Adventure|Sci-Fi


=== ratings_buffer ===
source: ../prepared/ratings_buffer.parquet
source: ../prepared/ratings_buffer.parquet


userId  movieId  rating            datetime               title  \
0    8780   164179     5.0 2018-01-04 02:45:08      Arrival (2016)   
1    8780   134130     5.0 2018-01-04 02:45:12  The Martian (2015)   
2    8780    70286     5.0 2018-01-04 02:45:17   District 9 (2009)   
3  115638    75341     4.0 2018-01-04 02:45:25  Remember Me (2010)   
4  115638     5564     4.0 2018-01-04 02:45:41      Swimfan (2002)   

                    genres  
0                   Sci-Fi  
1   Adventure|Drama|Sci-Fi  
2  Mystery|Sci-Fi|Thriller  
3            Drama|Romance  
4                 Thriller


=== movie_features_small ===
source: ../prepared/movie_features_small.parquet


movieId                               title  (no genres listed)  Action  \
0        1                    Toy Story (1995)                   0       0   
1        2                      Jumanji (1995)                   0       0   
2        3             Grumpier Old Men (1995)                   0       0   
3        4            Waiting to Exhale (1995)                   0       0   
4        5  Father of the Bride Part II (1995)                   0       0   

   Adventure  Animation  Children  Comedy  Crime  Documentary  ...  g_emb_54  \
0          1          1         1       1      0            0  ... -0.049991   
1          1          0         1       0      0            0  ...  0.080302   
2          0          0         0       1      0            0  ... -0.003513   
3          0          0         0       1      0            0  ... -0.316550   
4          0          0         0       1      0            0  ...  0.262982   

   g_emb_55  g_emb_56  g_emb_57  g_emb_58  g_emb_59  g_emb_60  g_emb_61  \
0  0.101698 -0.245650  0.004215 -0.133796  0.326354  0.073053  0.151898   
1  0.077730 -0.001964 -0.169890 -0.328397 -0.155135 -0.324006  0.229122   
2 -0.255320 -0.047668  0.043896  0.025104 -0.006311  0.120793  0.128007   
3  0.234972  0.107891  0.136243  0.067892  0.319507  0.008702  0.139247   
4 -0.265471  0.022237  0.152904  0.167361  0.148734  0.113982  0.046446   

   g_emb_62  g_emb_63  
0 -0.005448 -0.211587  
1  0.065407  0.075267  
2 -0.087302 -0.039096  
3 -0.150832  0.039255  
4 -0.158512  0.106765  

[5 rows x 86 columns]


=== movie_features_uni ===
source: ../prepared/movie_features_uni.parquet
source: ../prepared/movie_features_uni.parquet


movieId                               title  (no genres listed)  Action  \
0        1                    Toy Story (1995)                   0       0   
1        2                      Jumanji (1995)                   0       0   
2        3             Grumpier Old Men (1995)                   0       0   
3        4            Waiting to Exhale (1995)                   0       0   
4        5  Father of the Bride Part II (1995)                   0       0   

   Adventure  Animation  Children  Comedy  Crime  Documentary  ...     1119  \
0          1          1         1       1      0            0  ...  0.04050   
1          1          0         1       0      0            0  ...  0.05250   
2          0          0         0       1      0            0  ...  0.06275   
3          0          0         0       1      0            0  ...  0.05325   
4          0          0         0       1      0            0  ...  0.05350   

      1120     1121     1122     1123     1124     1125     1126     1127  \
0  0.01425  0.03050  0.03500  0.14125  0.05775  0.03900  0.02975  0.08475   
1  0.01575  0.01250  0.02000  0.12225  0.03275  0.02100  0.01100  0.10525   
2  0.01950  0.02225  0.02300  0.12200  0.03475  0.01700  0.01800  0.09100   
3  0.02800  0.01675  0.03875  0.18200  0.07050  0.01625  0.01425  0.08850   
4  0.02050  0.01425  0.02550  0.19225  0.02675  0.01625  0.01300  0.08700   

      1128  
0  0.02200  
1  0.01975  
2  0.01775  
3  0.01500  
4  0.01600  

[5 rows x 1150 columns]


=== movie_embeddings ===
source: ../prepared/movie_embeddings.parquet


movieId   g_emb_0   g_emb_1   g_emb_2   g_emb_3   g_emb_4   g_emb_5  \
0        1  5.691518  1.157255 -2.299418  0.432748  1.473947  1.740594   
1        2  3.842369  2.398972 -1.019800  0.010046  1.318503  0.216741   
2        3  2.881675  1.250844 -0.641894 -0.263696 -0.496165  0.352702   
3        4  2.781196  0.691665 -1.035578 -0.691427 -0.678009 -0.502189   
4        5  2.862799  1.388640 -1.272278 -0.620198 -0.514246  0.082957   

    g_emb_6   g_emb_7   g_emb_8  ...  g_emb_54  g_emb_55  g_emb_56  g_emb_57  \
0  0.675827  1.929924 -0.530852  ... -0.049991  0.101698 -0.245650  0.004215   
1  0.049389  1.165675  0.326293  ...  0.080302  0.077730 -0.001964 -0.169890   
2 -0.112911 -0.016496  0.080668  ... -0.003513 -0.255320 -0.047668  0.043896   
3 -0.735467  0.146575  0.280363  ... -0.316550  0.234972  0.107891  0.136243   
4 -0.343213  0.313056  0.034648  ...  0.262982 -0.265471  0.022237  0.152904   

   g_emb_58  g_emb_59  g_emb_60  g_emb_61  g_emb_62  g_emb_63  
0 -0.133796  0.326354  0.073053  0.151898 -0.005448 -0.211587  
1 -0.328397 -0.155135 -0.324006  0.229122  0.065407  0.075267  
2  0.025104 -0.006311  0.120793  0.128007 -0.087302 -0.039096  
3  0.067892  0.319507  0.008702  0.139247 -0.150832  0.039255  
4  0.167361  0.148734  0.113982  0.046446 -0.158512  0.106765  

[5 rows x 65 columns]


movie_features_small sample columns:


movieId                               title   g_emb_0   g_emb_1
0        1                    Toy Story (1995)  5.691518  1.157255
1        2                      Jumanji (1995)  3.842369  2.398972
2        3             Grumpier Old Men (1995)  2.881675  1.250844
3        4            Waiting to Exhale (1995)  2.781196  0.691665
4        5  Father of the Bride Part II (1995)  2.862799  1.388640

## Summary: Files for ML Model Training

### Overview
The preparation notebook has generated **clean, ready-to-use artifacts** for training baseline and hybrid recommendation models. All files are saved in the `prepared/` directory in both Parquet (preferred) and CSV formats.

---

### Core Artifacts

#### 1. **Interaction Data** (for baseline and hybrid models)
- **`ratings_train.parquet`** (70% of data, oldest)
  - Use as primary training set for all models (baseline + hybrid).
  - Columns: `userId`, `movieId`, `rating`, `timestamp`, `datetime`, `year`, `liked`, `title`, `genres`.
  - Size: ~17.5M rows.

- **`ratings_test.parquet`** (20% of data, middle)
  - Use for evaluation and hyperparameter tuning.
  - Same structure as training set.
  - Size: ~5M rows.

- **`ratings_buffer.parquet`** (10% of data, most recent)
  - Simulates new incoming data post-deployment.
  - Use for continuous monitoring, model drift detection, and retraining pipelines.
  - Size: ~2.5M rows.

---

#### 2. **Item Features** (for hybrid models)

**Recommended:** Use `movie_features_small.parquet` (optimized for training).

- **`movie_features_small.parquet`** ⭐
  - One row per movie.
  - Columns: `movieId`, `title`, genre one-hot columns (int8), PCA embeddings (`g_emb_0`...`g_emb_63`).
  - Size: ~62k rows × 86 cols (compact, memory-efficient).
  - Explained variance from PCA: ~90%+ of genome tag information retained.
  - **Use case:** Load into memory once; join by `movieId` at batch time in your dataloader.

- **`movie_embeddings.parquet`** (reference only)
  - Just the PCA-reduced genome embeddings keyed by `movieId`.
  - Useful if you want to replace genome features post-hoc or analyze embeddings.

- **`movie_features_uni.parquet`** (full features, reference)
  - One row per movie with genres + all 1,128 raw genome tag columns.
  - Use only for exploratory analysis; too large for direct training merges.

---

### How to Use for Training

#### **Baseline Model** (Collaborative Filtering)
```
1. Load ratings_train.parquet
2. Extract: userId, movieId, rating (or liked for binary classification)
3. Split further (e.g., train/val) if needed
4. Train CF model (SVD, ALS, NMF, etc.)
5. Evaluate on ratings_test.parquet
6. Monitor on ratings_buffer.parquet for drift
```

#### **Hybrid Model** (Collaborative + Content-Based)
```
1. Load ratings_train.parquet → interactions
2. Load movie_features_small.parquet → item features (keep in memory)
3. In your dataloader:
   a. Batch from ratings_train
   b. For each movieId, join from movie_features_small
   c. Concatenate: [userId_embedding, rating, movie_genres, g_emb_0...g_emb_63]
4. Train neural net (e.g., MLP, RNN) on enriched features
5. Evaluate on ratings_test.parquet with same join logic
6. Monitor on ratings_buffer.parquet
```

---

### Key Decisions Made

| Decision | Rationale |
|----------|-----------|
| **25M interactions (not 68M)** | Avoided row explosion by merging ratings with base movies (non-exploded genres). Exploded genres are available in `movie_features` for reference. |
| **Separate item features** | Prevented OOM: 25M × 1.1k tag cols would require ~110 GB. Instead, keep `movie_features_small` (~500 MB) in memory and join per-batch. |
| **PCA to 64 dims** | Reduced genome columns from 1,128 → 64 embeddings; retained ~90% variance; 5× smaller memory footprint. |
| **Temporal 70/20/10 split** | Ensures no time-leakage; buffer simulates production drift. |

---

### Next Steps for ML Training

1. **Load** `ratings_train.parquet` and `movie_features_small.parquet`.
2. **Baseline:** Train a simple CF model (SVD, for quick wins).
3. **Hybrid:** Add movie embeddings to baseline; compare performance.
4. **Tuning:** Experiment with embedding dimensions, regularization, etc.
5. **Monitoring:** Evaluate on `ratings_test.parquet`; detect drift on `ratings_buffer.parquet`.
6. **Deployment:** Set up re-training pipeline using recent buffer data.

---

### File Sizes (approx.)
- `ratings_train.parquet`: ~1.5 GB
- `ratings_test.parquet`: ~430 MB
- `ratings_buffer.parquet`: ~215 MB
- `movie_features_small.parquet`: ~500 MB
- **Total:** ~2.6 GB (easily fits on modern machines)

Good luck with model training! 🚀
